### PR TITLE
Require disability

### PR DESCRIPTION
**Pull request by cvalarida**
## Description
Requires a disability to be claimed. We're doing this in a few ways:

1. If the veteran is claiming increase only, we require them to select a rated disability
2. If the veteran is claiming new only, we require them to add a new disability
3. If the veteran is claiming both, we require they claim _either_ a rated _or_ a new disability
   - If the veteran did _not_ select a rated disability
     - We require them to select "Yes" to the "do you have any new conditions to claim" question
     - We require them to actually add a new disability

## Testing done
None yet

### Increase only
image

### New only
image

**Note:** I wasn't able to get the red line to show up for the validation error, so it's not featured in this PR. :disappointed: 

#### No selected
image

#### Disability not added
image

## Acceptance criteria
- [x] A user can't continue past the rated disability selection screen if they're claiming an increase only and have selected no conditions
- [x] A user can't continue past the yes/no question if they select "no" and haven't selected any rated disabilities
- [x] A user can't continue past the `add new disabilities` screen if
    - They're claiming both new and increase but didn't selected any rated disabilities nor added any new ones
    - OR they're claiming just new conditions but didn't add any
- [ ] The alert box is focused on when it shows up so screen readers read out the error (cc @1Copenut)

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Comment by peggygannon** [comment]
@cvalarida Here's a suggested message for "new " disability only: 

**We need you to add a disability**
You need to add a new disability to claim. We can't process your claim without a disability selected.

File: src/applications/disability-benefits/all-claims/content/addDisabilities.jsx [hunk 1]
```diff
@@ -27,3 +29,42 @@ export const uiDescription = (
...
 );
+
+const alertContent = newOnly => (
+  <>
+    <p>
+      You’ll need to add a new disability or choose a rated disability to claim.
+      We can’t process your claim without a disability selected. Please add a
+      new disability or choose a rated disability for increased compensation.
+    </p>
+    {!newOnly && (
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/content/addDisabilities.jsx]: @katelyncai should we have a link for the 'new only' alert that takes users back to the 'new or increase' selection page, or just leave it like this and omit the link in the 'new only' flow's alert box?

Comment (markgreenburg) [reply src/applications/disability-benefits/all-claims/content/addDisabilities.jsx]: 🤔 this also doesn't show when you select both new & rated options.

File: src/applications/disability-benefits/all-claims/pages/addDisabilities.js [hunk 2]
```diff
@@ -38,7 +51,25 @@ export const uiSchema = {
...
+      },
+    },
+  },
+  // This object only shows up when the user tries to continue without claiming either a rated or new condition
+  'view:newDisabilityErrors': {
+    'view:newOnlyAlert': {
+      'ui:description': newOnlyAlert,
+      'ui:options': {
+        hideIf: formData =>
+          !newConditionsOnly(formData) || hasClaimedConditions(formData),
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/pages/addDisabilities.js]: Are both of these needed? If they don't have any claimed conditions by this point, it probably doesn't matter whether or not they selected something other than 'new only,' I'd think we just want them to enter a new disability regardless, right?

File: src/applications/disability-benefits/all-claims/pages/addDisabilities.js [hunk 3]
```diff
@@ -38,7 +51,25 @@ export const uiSchema = {
...
+      },
+    },
+    'view:increaseAndNewAlert': {
+      'ui:description': increaseAndNewAlert,
+      'ui:options': {
+        hideIf: formData =>
+          // Only show this alert if the veteran is claiming both rated and new conditions
+          !newAndIncrease(formData) || hasClaimedConditions(formData),
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/pages/addDisabilities.js]: 🤔 if we want this to show to veterans that *are* claiming 'new and increase', is `!newAndIncrease` what we want?

**Comment by 1Copenut** [comment]
I tested this behavior in Chrome and Safari with VoiceOver. The alert box fires as expected, and is read aloud. However, VoiceOver doesn't read the entire alert box message, it's too long. Because we are moving focus back to the form element(s) in error, users don't ever get the chance to hear the entire alert message read aloud. 

I think we have a couple of options:

1. We go against pattern and explicitly set focus on the alert box instead of the form input. I have reservations about this because it breaks our normal, expected behavior when forms have errors.
2. We shorten the messages in our alert boxes to be read out in one announcement by the screen readers. This would require additional testing, because each reader has its own limits for characters in memory. 

I'm happy to discuss this further tomorrow.

**Comment by cvalarida** [comment]
@1Copenut I think this is what's happening:
1) The user hits "Continue" to trigger the error message
2) The normal validation message (tied to the input field) and the alert box are rendered at the same time
3) Because the alert box has `aria-live="polite"`, it gets read out (or started to, anyhow)
4) The focus is then deliberately shifted to the first input error, so _that_ gets read out

I tried deliberately focusing on the alert box, but without doing some `setTimeout` or weird event firing, I wasn't able to make it happen because the error message is focused on _after_ the alert box's `componentDidMount` fires.

I've got some reservations about both options you propose, so I'm not sure what the "right" solution is here.
1) I could use a `setTimeout` or potentially even some fancy event firing to focus on the alert box, but I'm afraid it'll read like
   1) Alert box header
   2) Validation message (in part or in full, I don't know)
   3) Alert box header and content
2) Shortening the alert box messages is maybe feasible, but relying on something to finish before something else is...basically always buggy
   - To be more specific, screen reader speed is variable; if it's set to a slower speed, it may _still_ not finish before we refocus on the input

**Review by markgreenburg — COMMENTED** [review]
Looks good, but couple of things that I'm not sure about:
1. The alert box messaging for different scenarios seems like it might be off a bit. E.g., for the 'New Only' flow, the alert message mentions 'select a rated disability', which isn't possible unless we add a link to go back to the path selector page (new vs. increase vs. both)
2. In the 'new only' flow, we still show the 'Do you want to add new conditions' page, which seems duplicative. I'd think that we wouldn't want to keep this page in that flow.

So technically I think this is fine, but I'd like to circle back with Peggy and Katelyn / Alex on the content and UX before we merge this in.

**Comment by 1Copenut** [comment]
@cvalarida I should clarify my point about the screen reader and shortening the alert box. Screen readers load the next N number of characters into memory, and read until the message is complete, or the character limit is reached. Users advance the virtual cursor at that point to hear the rest of the paragraph or message.

Where we're moving focus back to the radio buttons or checkboxes, I'm afraid screen reader users will not hear the entire alert box read aloud unless we make the message shorter.

**Comment by katelyncai** [comment]
Ok, I'm trying to play catch-up on all the good conversation here.
One question I have: do we know how many characters are loaded in the memory? Is that something set by individuals? Is there a default we can strive to stay under?

Agreed @markgreenburg, the messaging for `New Only` should reflect that you need to add a new disability. I am bit lost with where we stand with each message. The original intention for the longer alert box was to make the copy generic enough so that it could work with each scenario. @goldenmeanie and I thought it couldn't hurt to explain that you need one or the other.

**Comment by 1Copenut** [comment]
@katelyncai && @goldenmeanie 
The number of characters loaded into SR memory is dependent on the software. JAWS stores somewhere around 80 IIRC. VoiceOver reads out the alertbox heading and the first three lines of text, give or take a few. Not sure about NVDA without testing. Since users have interacted with the page by the time they see the error message, we can't count on the auto-read feature. 

Pressing `Continue` automatically triggers the alertbox, then moves focus to the input that is causing the error. I don't think it's a good idea to try to override that flow, or move focus a second time to settle on the alertbox. This could have unpredictable results for screen readers and keyboard navigation.

I could see either shorter alert box messages, or a shorter message being added to the form input's regular error (instead of the alertbox) being ways around this issue, but can appreciate there might not be time for refactoring on a first release. I don't consider this a launch blocker, but would like to figure out a long-term solution that makes for a great UX and is fully compliant too.

Let me know if you want to discuss this further on Slack or video chat.

**Comment by cvalarida** [comment]
@goldenmeanie @katelyncai Does it still make sense to have this page if the user selected "new only" for the claim type?
image

**Comment by goldenmeanie** [comment]
@cvalarida I think this screen should be skipped if they said they have new conditions earlier.

@1Copenut How would this play out if we appended the alert to the input label?

**Comment by 1Copenut** [comment]
@goldenmeanie I envisioned a more detailed error message that could say "Alert! You need to add a new disability or choose a rated disability to claim", instead of the alertbox, as one solution. That eliminates the tricky focus, then focus again logic, but removes a strong visual element in the alertbox. I'm not sure I'm doing a good job explaining in comments; let me know if you'd like to review in a quick video chat or schedule something with engineering too.

**Review by peggygannon — APPROVED** [review]
@cvalarida looks good to me...

**Comment by cvalarida** [comment]
I think there's more that needs to be done to hide that page, so I've pulled it out into [this ticket](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/17284).

*Note:* This will introduce an edge case bug when:
1) The user selects the new-only path
2) They hit "no" for the yes/no question
3) They follow the link provided to take them to the rated disabilities page
Result: They'll get directed back to the intro page and will have to hit the "continue" button again.

**Review by markgreenburg — APPROVED** [review]
LGTM with the understanding that the related ticket (https://github.com/department-of-veterans-affairs/vets.gov-team/issues/17284) is addressed as a quick follow to fix the edge-case bug with certain alert box links.

**Pull request merged**